### DISCUSSION TITLE
Roughnator and Kitt4sme.live comms

### DISCUSSION THREAD

**Discussion by berserkrambo** [c0c0n3/kitt4sme.live#183] [Q&A] [answered]
Hi, I'm Riccardo from AI4SDW,
I have a question about the communication between roughnator and kitt4sme.live. Is it doable?

For what I understood, the Roughnator simulation actually "emulates" kitt4sme.live by running every module he needs through a docker-compose env.
But there is a way to utilize the kitt4sme.live local instance instead of using the one inside the Roughnator repository? And if yes, how can I check that this works properly?

For example:
Let's suppose to have kitt4sme.live running in a multipass instance with this ip addr: x.x.y.y and we are able to see argocd via browser at x.x.y.y/argocd, also all services are in the "running" status (visible through kubectl get pod --all-namespaces).

In the roughnator docker-compose we comment:
#-mongo
#-orion
#-crate
#-quantumleap
#-dazzler

and we change ip add in tests/util/fiware.py
ORION_EXTERNAL_BASE_URL = 'http://localhost:1026/' -->> ORION_EXTERNAL_BASE_URL = 'http://x.x.y.y:1026/' (multipass instance ip addrr)
ROUGHNATOR_INTERNAL_BASE_URL = 'http://roughnator:8000/' -->> not sure what this is for
QUANTUMLEAP_INTERNAL_BASE_URL = 'http://quantumleap:8668/' -->> not sure what this is for
QUANTUMLEAP_EXTERNAL_BASE_URL = 'http://localhost:8668/' -->> QUANTUMLEAP_EXTERNAL_BASE_URL = 'http://x.x.y.y:8668/' (multipass instance ip addrr)

Shouldn't this work and make roughnator communicate with my local kitt4sme.live? I tried this, but it didn't work. Maybe because vms belonging to the docker compose env see only "simtests" network? What's the right way to make my local roughnator communicate with my local kitt4sme.live?

The right way to integrate our services with kitt4sme could be to "copy" roughnator and replace it with our AI lib (in case of an AI service) and do the same with dazzler in case of an UI service?
Then dockerize and put them together in a docker compose env (like roughnator) and check if all this works?
If this is the way to go, how am I supposed to use the kitt4sme.live?

Thank you in advance for your time.

Bests,
Riccardo

**Comment by c0c0n3** [comment, marked as answer]
Hi @berserkrambo :-)

> communication between roughnator and kitt4sme.live. Is it doable?

yep, Roughnator is actually live in our reference cloud instance at kitt4sme.collab-cloud.eu.

> the Roughnator simulation actually "emulates" kitt4sme.live by running every module he needs through a docker-compose env.

Correct. Well, I'm assuming you're referring to this:
- https://github.com/c0c0n3/kitt4sme.roughnator#live-simulator
- https://github.com/c0c0n3/kitt4sme.roughnator/tree/main/tests/sim

> way to utilize the kitt4sme.live local instance instead of using the one inside the Roughnator repository? 

If you use our reference implementation repo (https://github.com/c0c0n3/kitt4sme.live) to instantiate a kitt4sme cluster on your own hardware following the bootstrap procedure, then Roughnator will be in your cluster:

- https://github.com/c0c0n3/kitt4sme.live/blob/main/deployment/plat-app-services/kustomization.yaml#L9

> how can I check that this works properly?

Either log into Argo CD (http://your-local-kitt4sme-cluster/argocd) and click on the Roughnator app to see the details or do that from the command line, e.g. `kubectl get pod`, `kubectl logs svc/roughnator`, etc. Also keep in mind you can see traffic in real time through our Kiali dashboard `istioctl d kiali`.

> For example:
>  kitt4sme.live running in a multipass instance with this ip addr: x.x.y.y
> able to see argocd via browser at x.x.y.y/argocd, 
> also all services are in the "running" status (visible through kubectl get pod --all-namespaces).

Then if you instantiated your Multipass VM following the bootstrap procedure,  you should be able to see Roughnator running inside the K8s which is inside your VM---Russian dolls, kinda mental, I know :-)

> In the roughnator docker-compose we comment:

Not sure why you'd use docker compose in this scenario, since Roughnator along with all the other services in that docker compose file are running inside your Multipass VM. In other words, your setup is pretty much what you see in this picture

- https://raw.githubusercontent.com/c0c0n3/kitt4sme.roughnator/main/roughnator-sim.svg

except what's inside the "docker compose" dashed rectangle in that picture now sits inside your Multipass VM.

> Shouldn't this work and make roughnator communicate with my local kitt4sme.live?

Do you mean shouldn't the simulation work as explained on the Roughnator home page? I never tested this scenario, but it might actually work with the env mods you mentioned. Of course, the service you should look at though is the Roughnator instance inside your VM not that in docker compose. Actually if you go down this road, i.e. use the simulator driver to set up subscriptions inside the VM and send data from localhost to the VM, then I'd get rid of docker compose since it serves no purpose. Just comment out the following lines in the simulator driver

- https://github.com/c0c0n3/kitt4sme.roughnator/blob/main/tests/sim/runner.py#L11-L12
- https://github.com/c0c0n3/kitt4sme.roughnator/blob/main/tests/sim/runner.py#L31

> to integrate our services with kitt4sme

you should fork https://github.com/c0c0n3/kitt4sme.live and instantiate your own kitt4sme cluster from your fork---see the [bootstrap procedure](https://github.com/c0c0n3/kitt4sme.live/blob/main/docs/bootstrap.md) for the details. Then put together some basic K8s deployment for your services---AI, UI, whatever you've got. Commit those to your fork and Argo CD in your cluster should pick up the new files and deploy them---see [how to deploy an app](https://github.com/c0c0n3/kitt4sme.live/blob/main/docs/how-to-deploy-app-svc.md). Keep in mind our GitOps pipeline basically runs just plain K8s/Kustomize so you can just as well deploy from the command line with e.g. `kustomize build deployment/plat-app-services/my-app | kubectl apply -f -`. That's more or less what Argo CD does under the bonnet for you.

> Then dockerize and put them together in a docker compose env (like roughnator) and check if all this works?

Well, you've got to have Docker images before you can deploy stuff to a K8s cluster. But yea, after you've put together an image, then you could also whip together a simple Docker compose to test your setup before venturing in K8s land. That's what we've done for all our AI services---the Roughnator simulator is a good case in point. In fact, it's much easier to test your services work as expected in a simpler env like Docker compose, less moving parts, less things that can go wrong. After you have a solid baseline, you could tackle the actual cloud deployment. That's just a suggestion obviously, follow whatever approach works best for you :-)

One last thing. If you're doing Python, you could actually lift the dev env, docker and CI/CD pipeline setup from Roughnator. Tweak the poetry stuff to match your dev env, the docker file to build your image and supposedly you could keep the content of the `.github` dir  as it is. Then every time you release, the GitHub CI/CD pipeline should automatically build and publish your docker image.

Hope this helps!

**Reply by berserkrambo** [reply]
Hi @c0c0n3 and happy new year! :)

I have a new question, it's about the communication between the external world and the kitt4sme.live platform.

I followed your suggestion:

>Well, you've got to have Docker images before you can deploy stuff to a K8s cluster. But yea, after you've put together an image, then you could also whip together a simple Docker compose to test your setup before venturing in K8s land. That's what we've done for all our AI services---the Roughnator simulator is a good case in point. In fact, it's much easier to test your services work as expected in a simpler env like Docker compose, less moving parts, less things that can go wrong. After you have a solid baseline, you could tackle the actual cloud deployment. That's just a suggestion obviously, follow whatever approach works best for you :-)


So.. I did that and now I have our service and our dashboard running inside our kitt4sme.live fork inside a multipass virtual machine and I see them correctly through the argocd web app.

And now? How can I check if they actually works? I mean, I know they works because I tested the entire pipeline in a docker compose env but... how can I test/ping/communicate with/send some data to them while they are inside the kitt4sme.live platform?
What are the next steps?

Thanks! :)

**Reply by kostasgrevenitis** [reply]
Good morning and happy new year. 

Have you assigned to your VM any public IPs? if yes, then send data to the platform by using those public IPs. You can use POSTman or another similar tool.

**Reply by berserkrambo** [reply]
Hi, thank you for the quick answer.
Ok, let's say my multipass vm has this ip: 10.140.106.105

I'll try to send some data to orion using postman or using your python ngsy script editing the `ORION_EXTERNAL_BASE_URL`

Question:
To access argocd I can simply do it via web browser from the host machine navigating to 10.140.106.105/argocd.
Can I access to our dashboard the same way? e.g. 10.140.106.105/k4sme-ai4sdw-dash?  (k4sme-ai4sdw-dash is the service name of our dash and...I tried and it does not work xD)

thank you

**Reply by kostasgrevenitis** [reply]
Good morning.

Check your dashboard yaml file. Maybe a configuration about the port or the host is missing?

**Comment by c0c0n3** [comment]
Happy 2023 @berserkrambo!

> Can I access to our dashboard the same way? e.g. 10.140.106.105/k4sme-ai4sdw-dash

Yes, but you need to configure URL-based routing in Istio. Here's how that's done for ArgoCD

- https://github.com/c0c0n3/kitt4sme.live/blob/main/deployment/mesh-infra/routing/ingress.yaml#L60-L71

You'll have to add a similar `match` stanza for your service.

As usual, with K8s and Istio there's more than one way to skin a cat. So you could instead set up port-based routing, or expose a node port, or route based on HTTP headers, etc. If you've never done this kind of thing before, maybe have a quick read through the traffic management/routing sections of K8s & Istio's manuals to see what's possible, then take it from there.

**Comment by berserkrambo** [comment]
Hi @c0c0n3 and @kostasgrevenitis,

I'm still having trouble seeing my dashboard@kitt4sme.live cluster...


I added new match in the `kitt4sme.live/blob/main/deployment/mesh-infra/routing/ingress.yaml`:
```
- match:
      - uri:
          prefix: /k4sme-ai4sdw-dash/
      - uri:
          prefix: /k4sme-ai4sdw-dash
      rewrite:
        uri: /
      route:
        - destination:
            host: k4sme-ai4sdw-dash.default.svc.cluster.local
            port:
              number: 8000
```

I also added a rule in the `kitt4sme.live/deployment/mesh-infra/security/ingress-policy.yaml`:
```
  rules:
  - to:
    - operation:
        # Ask OPA for permission to go ahead w/ any incoming request except
        # for requests to Keycloak, Argo CD or to retrieve FIWARE service
        # versions where there's no need for authorisation.
        notPaths:
        - "/auth"
        - "/auth/*"
        - "/argocd"
        - "/argocd/*"
        - "/dazzler/*"
        - "/orion/*"
        - "/orion/version"
        - "/quantumleap/version"
        - "/sads/*"
        - "/k4sme-ai4sdw-dash/*"
        - "/k4sme-ai4sdw-dash"
```

But.... when I try to access my dashboard at http://10.140.106.105/k4sme-ai4sdw-dash I get a beautiful 404 error..:(
My services are up and running, I see them correctly through `kubectl get pod --all-namespaces` and `kubectl get pod --all-namespaces`

```
default        pod/k4sme-ai4sdw-dash-796c879784-2qtkp           2/2     Running    12         22d
default        pod/k4sme-ai4sdw-ai-7c8589db46-z2mqr             2/2     Running    12         22d
default        service/k4sme-ai4sdw-dash           ClusterIP      10.152.183.12    <none>        8000/TCP                                                               22d
default        service/k4sme-ai4sdw-ai             ClusterIP      10.152.183.133   <none>        8000/TCP                                                               22d
default        deployment.apps/k4sme-ai4sdw-dash           1/1     1            1           22d
default        deployment.apps/k4sme-ai4sdw-ai             1/1     1            1           22d
default        replicaset.apps/k4sme-ai4sdw-dash-796c879784           1         1         1       22d
default        replicaset.apps/k4sme-ai4sdw-ai-7c8589db46             1         1         1       22d
```

What am I doing wrong? there's something else to change somewhere?

p.s. my repos are public, you can see my ai service at https://github.com/berserkrambo/k4sme.ai4sdw.ai and my dashboard at https://github.com/berserkrambo/k4sme.ai4sdw.dash

p.p.s if I send some data to orion using `MachineSampler` Roughnator service seems to work fine but if I try to access dazzler http://10.140.106.105/dazzler/demo/-/roughnator I get no dashboard, instead I get a `{"detail":"Not Found"}` message...
Also my ai service seems to work fine. If I send data to orion using my tenant's tag `ai4sdw`, orion triggers and send data to my ai service...I just can't access my dashboard via browser....

thanks :)

**Reply by c0c0n3** [reply]
Hi @berserkrambo!

It looks like you got quite far along w/ the integration, good stuff. Let's see if we can go past the last hurdles before the finish line.

> access my dashboard at http://10.140.106.105/k4sme-ai4sdw-dash I get a beautiful 404 error

Holy moly :-) It looks like your Istio routing is fine, so most likely it's actually the dashboard service returning that 404. With URL-based routing, that typically happens b/c the service is expecting to serve HTTP resources at a root of `/` whereas in your case the URL root is `/k4sme-ai4sdw-dash`. Long story short: I think all you need to do is configure the dashboard to use a URL root of `/k4sme-ai4sdw-dash` instead of the default (?) of `/`.

Just looked at your code, your setup is pretty similar to our Flaw Sleuth (officially known as SADS :-)
- https://github.com/c0c0n3/kitt4sme.flaw-sleuth

So you should be able to tell Streamlit to use `/k4sme-ai4sdw-dash` as base URL, e.g. in your K8s deployment

```yaml
         env:
          - name: "STREAMLIT_SERVER_BASE_URL_PATH"
            value: "/k4sme-ai4sdw-dash"
```

You can have a look at this PR which deployed Flaw Sleuth to our live cluster, your deployment should most likely be similar
- https://github.com/c0c0n3/kitt4sme.live/pull/159

also I think you've got to start Streamlit in headless mode, see what we did here in Flaw Sleuth's Docker file:
- https://github.com/c0c0n3/kitt4sme.flaw-sleuth/blob/main/Dockerfile#L19

Hope this helps!

**Reply by berserkrambo** [reply]
Hi @c0c0n3,

I followed your advices (check at https://github.com/berserkrambo/k4sme.ai4sdw.dash/blob/main/Dockerfile#:~:text=COPY%20streamlit%2Dapp,app.py%22%5D and the routing rules at https://github.com/berserkrambo/kitt4sme.live/blob/main/deployment/mesh-infra/routing/ingress.yaml#:~:text=%2D%20match%3A%20%23,number%3A%208501). 
Now the settings of my dashboard are almost identical to those of your sads but... I'm still not able to see my dashboard via browser at http://10.140.106.105/k4sme-ai4sdw-dash/.

Just for fun I replaced your sads docker image with the one of my dashboard and..tadaaaaa, now at http://10.140.106.105/sads I can finally see my (not) beautiful dashboard so... the problem has to be somewhere inside the routing rules I think…

I don't know what to do anymore...maybe ingress doesn't like the name with two dash? xD It'like ingress doesn't trigger when I ask for the 'k4sme-ai4sdw-dash' page, I get a 404 error as if I was going to a random "hello" page...

Thanks for helping us :)

**Reply by berserkrambo** [reply]
And here I am to reply to myself, it really was a problem relative to ingress policies, specifically it was a f...... bad indentation:
from
```
  - match: # NOTE (4)
      - uri:
``` 
to
```
  - match:  # NOTE (4)
    - uri:
```
really life changing lol

thanks again :)

**Reply by c0c0n3** [reply]
@berserkrambo ha! good catch! Welcome to the cloud computing YAML madness. I wish DevOps valued more type theory...but never mind. Anyway for the record I tested that too and it works fine on my side. After building a test VM as explained in the bootstrap procedure, I checked out your kitt4sme.live fork and then ran these commands

```bash
$ pwd
/Users/andrea/berserkrambo/kitt4sme.live
$ emacs k8s.yaml
$ export KUBECONFIG=/Users/andrea/berserkrambo/kitt4sme.live/k8s.yaml
$ cd deployment/
$ istioctl install -y --verify -f mesh-infra/istio/profile.yaml
$ kubectl label namespace default istio-injection=enabled
$ kustomize build mesh-infra/storage | kubectl apply -f -
$ kustomize build mesh-infra/routing | kubectl apply -f -
$ kustomize build mesh-infra/security/ | kubectl apply -f -
$ kustomize build plat-app-services/k4sme-ai4sdw-dash | kubectl apply -f -
```

then broswed to http://192.168.64.20/k4sme-ai4sdw-dash/, which correctly loaded your dashboard.

Happy days!

**Reply by berserkrambo** [reply]
Hi, @c0c0n3 , last question, I promise :)
Now that our system appears to work properly on our local instance kitt4sme.live, I'm not sure what are the steps to integrate it into your kitt4sme cloud instance. Did I miss any steps while reading the documentation? Can you give me some advice?
Thanks!

**Comment by c0c0n3** [comment]
Hi @berserkrambo :-)

> system appears to work properly on our local instance kitt4sme.live

great job!

> what are the steps to integrate it into your kitt4sme cloud instance

I don't think you'll have to do that. My understanding is that the open call evaluators are going to look at your local instance as integration evidence? In other words, you're done. But please double-check that with your mentor, I'm not involved with open calls so I could be wrong...

**Reply by berserkrambo** [reply]
Ah ok, top! we'll check with our mentor. 
Thank you!
